_SPIKE_.md

```diff
@@ -0,0 +1,543 @@
+
+## Current status
+
+UNIQUE index is supported as a specialized version of a generic index feature.
+
+Support for indexes and foreign keys depends on the connector. Some do support
+all features, some support only some of them. Depending on the database used,
+certain features are not possible to implement at all because of database design
+constraints. For example, CouchDB/Cloudant do not support UNIQUE indexes
+
+We have a syntax for defining (unique) indexes described in
+[our docs](https://loopback.io/doc/en/lb3/Model-definition-JSON-file.html#indexes),
+this syntax supports composite indexes. However, many connectors are adding
+their own extensions and PostgreSQL uses a slightly different format violating
+our docs.
+
+We have an undocumented syntax for defining foreign keys which is recognized by
+all SQL connectors that provide migration of foreign keys. Unfortunately, this
+syntax does not support composite foreign keys and there is also no support for
+expressing `ON UPDATE` and `ON DELETE` behavior.
+
+You can find more details in
+[Detailed description of the current syntax](#detailed-description-of-the-current-syntax)
+at the bottom.
+
+## Proposal
+
+### Indexes at property level
+
+Support the following two short-hand forms only. Ask users to use model-level
+form to define indexes that are more complex.
+
+- a "plain" index with no special configuration
+
+  ```js
+  {
+    email: {
+      type: 'string',
+      index: true,
+    }
+  }
+  ```
+
+- UNIQUE index with no special configuration
+
+  ```js
+  {
+    email: {
+      type: 'string',
+      unique: true,
+    }
+  }
+  ```
+
+We should also allow models to configure indexes for certain databases only by
+using database-specific options we already use to control column name, storage
+data type, etc.
+
+```js
+{
+  email: {
+    type: 'string',
+    mysql: {
+      index: true,
+      // unique: true
+    }
+  }
+}
+```
+
+### Indexes at model level
+
+Problem 1: different databases provide different indexing features.
+
+- It seems that all databases support ASC/DESC option at column-level.
+- MongoDB adds extra column-level option `text` as an alternative to ASC/DESC.
+- PostgreSQL allows column expressions, e.g. `lower(email)`.
+
+Problem 2: do we want to use database column names or LB property names? These
+two names can be different, especially when using database-specific config to
+map property names to custom column names.
+
+**The proposal:**
+
+At high-level, keep the current syntax where indexes are defined via a key-value
+map stored in `settings.indexes` property, the key is the index name and the
+value is an index definition object.
+
+```js
+{
+  strict: false,
+  forceID: true,
+  indexes: {
+    uniqueEmail: {
+      // index definition
+    },
+    nameQueries: {
+      // index definition
+    }
+  }
+}
+```
+
+Individual indexes can be defined as follows:
+
+- Add a new field `properties` as a key-value map from property names to
```

Comment:
I don't quite follow the meaning of "Individual indexes", could you explain more about it?

I understand the example above uses LB3 syntax:

```
{
  strict: false,
  forceID: true,
  indexes: {
    uniqueEmail: {
      // index definition
    },
    nameQueries: {
      // index definition
    }
  }
}
```
While confused about the individual index...is it a syntax supported in LB3 or a proposal that will be supported?

Reply:
I have same comment as @jannyHou .

Do you mean 

```
uniqueEmail: {
      // index definition
    },
```

can be filled in like this?: 

```
uniqueEmail: {
      properties: {
           email: 1, // ASC
          createdAt: 'DESC', // alias for -1
          bio: 'text', // database-specific value (MongoDB's "text")
       }
    },
```